### DISCUSSION TITLE
theme.css taking time to load

### DISCUSSION THREAD

**Discussion by imrankhurram-checkit** [primefaces/community#1779] [PrimeFaces] [answered]
I am using primefaces 12 and the theme is default theme saga. theme.css file is taking alot of time to load, it is the heaviest file in the website. it is taking 1.29, 1.45 and sometimes even more time to load. is there a way to make it fast?

**Comment by melloware** [comment]
I just went to the PF showcase which uses SAGA and presses CTRL+F5 to force a reload and it took 258 ms?

![image](https://github.com/primefaces/community/assets/4399574/989c8834-39b2-4e27-b6b1-8959d1750440)

Are you sure its not your network?  Is it that slow locally running on localhost?

**Comment by imrankhurram-checkit** [comment]
localhost is also having same results. Actually all the resources and the page itself is taking less time, only theme.css is taking too much time

**Reply by melloware** [reply]
I can't explain that it makes no sense. Can you try this sample.
[pf-11824.zip](https://github.com/primefaces/community/files/15142670/pf-11824.zip)

Unzip and run `mvn clean jetty:run` and navigate to http://localhost:8080  It uses `saga` and loads fast for me.

**Reply by imrankhurram-checkit** [reply]
mine has taken 1.25s just now on localhost

**Reply by imrankhurram-checkit** [reply]
ok may be it is not taking that much time for you. but is there a way to make it fast for me?

**Reply by melloware** [reply]
I think you have to get to the root of the problem of why. That sample I sent you takes less than 100ms for me.  Why your machine is taking 1.25 seconds seems crazy.  What are the specs of your machine?

**Reply by imrankhurram-checkit** [reply]
corei7 10th Gen. 16GB RAM. Also the trouble is we have tried it on multiple machines. Production website on production server is having that issue.
I know it is crazy, this is the problem?
I am using Wildfly30 and Primefaces 12

Is there a way to make it load fast?

**Reply by imrankhurram-checkit** [reply]
It is same on different browsers. Tried on Chrome and Mozilla

**Reply by melloware** [reply]
I am using Wildfly 24-31 with PF 10, 11, 12 ,13 all with lightning fast loads for themes.  

Are you saying my Jetty run is also slow for you?  That eliminates Wildfly as the culprit.

**Reply by imrankhurram-checkit** [reply]
yes your sample is taking less time but mine is not. one difference I can spot is your sample is using Javax whereas I am using Jakarta EE 10

**Reply by melloware** [reply]
Here is a similar test using Jakarta EE10 and Mojarra 4.x which is what Wildfly uses.
[pf-11578-jakarta.zip](https://github.com/primefaces/community/files/15148528/pf-11578-jakarta.zip)

**Reply by imrankhurram-checkit** [reply]
still not helping. I tried matching configs in web.xml but the theme.css is still taking too much time to load

**Reply by imrankhurram-checkit** [reply]
I am already using gzip compression which is also showing in developer tools that it is working. but still the file is taking time

**Reply by melloware** [reply]
Has to be something with your wildfly configuration any chance you could put together a small WAR file that I can deploy on Wildfly to show the issue?  As I said I use Wildfly 24-31 without issue.

**Reply by imrankhurram-checkit** [reply]
I have created a minimalistic project and the WAR is attached. On localhost it is taking 800+ ms which is alot among all other files.
![theme loading locally](https://github.com/primefaces/community/assets/64078961/7894cab1-d147-4733-9032-25db0267b944)

And when deployed on a remote machine it is taking 3+ seconds which is again alot as this is minimalistic project and not loading anything much and still one file taking alot of time.
![image](https://github.com/primefaces/community/assets/64078961/94a603f6-599c-472a-98ba-17aaf2bc55d1)

Unzip the file and check the WAR

[TestProjectV1.0.zip](https://github.com/primefaces/community/files/15296195/TestProjectV1.0.zip)

**Reply by melloware** [reply]
Oh its NOT just your theme.css.  Look at `primeicons.css` its only 3.6KB yet your browser says its taking 4.23 seconds to load.   I really think something is wrong with your setup / network

**Reply by imrankhurram-checkit** [reply]
please can you have a look, I am struggling alot with this. I am providing this minimalistic app here:
[TestProject-Mini.zip](https://github.com/primefaces/community/files/15296378/TestProject-Mini.zip)

**Reply by melloware** [reply]
Here is the primefaces Showcase: https://www.primefaces.org/showcase/index.xhtml?jfwid=dc968

Saga took 56 millseconds to load.

![image](https://github.com/primefaces/community/assets/4399574/23818a51-4141-4df5-8f05-b4a9543ed1f4)

**Reply by melloware** [reply]
I already see 1 problem why is your JSF Stage Development.  Your server will run much slower in Development mode.

```xml
<context-param>
		<param-name>jakarta.faces.PROJECT_STAGE</param-name>
		<param-value>Development</param-value>
	</context-param>
```

Also are you sure these other libs like PrettyFaces aren't slowing it down...

```xml
		<dependency>
			<groupId>org.ocpsoft.rewrite</groupId>
			<artifactId>rewrite-config-prettyfaces</artifactId>
			<version>10.0.1.Final</version>
		</dependency>
```

**Reply by imrankhurram-checkit** [reply]
I have already tried it in production mode as well. 
Can try removing pretty faces but this is the requirement to use it. I need to rename the URLs. Have you tried after removing it?

**Reply by melloware** [reply]
No I am asking you to debug.  You have to slowly narrow it down until you figure it out. I am not having the issue in any environment.

**Reply by imrankhurram-checkit** [reply]
doing both changes has not reduced the load time but it started putting this CSS file into cache properly which is making it fast. I'll have to check if that is acceptable in real app. 
Is there any other way to rewrite URLs which is fast and easy?

**Reply by melloware** [reply]
So are you using PrettyFaces just to get rid of the extensions like .jsf or .xhtml?  What is your use case?

and yes if the Cache headers were telling it not to cache of course your app will be slow loading on every page.  Here is a good explanation of Cache Control from OmniFaces: https://showcase.omnifaces.org/filters/CacheControlFilter

**Comment by OBUTTERLIN** [comment, marked as answer]
Are you running behind an Apache server ?
We had some issues with pages, not anytime, but with huge delta (up to 8s).
I think the problem is not in PrimeFaces or Wildfly or Jakarta, I'm using PrimeFaces 14 on Wildfly 31 and Jakarta EE10 and it take about 100ms to download.
